### PR TITLE
8190753: (zipfs): Accessing a large entry (> 2^31 bytes) leads to a negative initial size for ByteArrayOutputStream

### PR DESCRIPTION
**Pull request by apavlyutkin**
Hi!

Please review the backport of JDK-8190753 to 8u. The patch fixes IllegalArgumentException that takes place on accessing large (>2^31 bytes) entries in zipfs.

The original patch

https://github.com/openjdk/jdk11u-dev/pull/545

was applied with the following changes:

- changes to absent DeflatingEntryOutputStream class were skipped (in 8u EntryOutputStream handles both RAW and deflating streaming);
- the tests were updated to not use unsupported APIs like Path.of() and Map.of()
- timeout value for ZipFSOutputStreamTest was increased because the test is too greedy and may fail by timeout on low performance hosts

The changes were previously reviewed: https://mail.openjdk.java.net/pipermail/jdk8u-dev/2021-December/014479.html

Verification/regression: jdk_other (20.04/amd64)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8190753](https://bugs.openjdk.java.net/browse/JDK-8190753): (zipfs): Accessing a large entry (> 2^31 bytes) leads to a negative initial size for ByteArrayOutputStream

### Reviewers
 * [Paul Hohensee](https://openjdk.java.net/census#phh) (@phohensee - **Reviewer**) ⚠️ Review applies to 4ea72656882a84283ad4844385f1e1689f427c2f
 * [Andrew John Hughes](https://openjdk.java.net/census#andrew) (@gnu-andrew - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk8u-dev pull/17/head:pull/17` \
`$ git checkout pull/17`

Update a local copy of the PR: \
`$ git checkout pull/17` \
`$ git pull https://git.openjdk.java.net/jdk8u-dev pull/17/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17`

View PR using the GUI difftool: \
`$ git pr show -t 17`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk8u-dev/pull/17.diff">https://git.openjdk.java.net/jdk8u-dev/pull/17.diff</a>

</details>

### CODE REVIEW

**Review by phohensee — APPROVED** [review]
Lgtm.

**Comment by gnu-andrew** [comment]
Sorry for the delay on sorting this one out. It seems to have been bounced across all three repositories during the Git transition.

It now looks correct to me from a technical perspective. However, I am seeing some very odd whitespace differences between the two new test files when compared with the versions in 11u (either from the patch or the 11u-dev repository) e.g

```diff
-  @AfterMethod
+    @AfterMethod
     public void tearDown() throws IOException {
-    deleteFiles();
-  }
+        deleteFiles();
+    }
```

I don't know how this has occurred - I don't recall it being present in the original review - but I think the simplest way to fix it would be to store a diff between your 8u versions and the 11u versions using `-b`, replace the 8u ones with the ones from 11u, and then reapply your changes for 8u from the diff i.e. something like

```shell
$ diff -ub jdk/test/demo/zipfs/ZipFSOutputStreamTest.java ../jdk11u-dev/test/jdk/jdk/nio/zipfs/ZipFSOutputStreamTest.java > /tmp/test1.patch
$ cp ../jdk11u-dev/test/jdk/jdk/nio/zipfs/ZipFSOutputStreamTest.java jdk/test/demo/zipfs
$ <edit test1.patch to remove any remaining unneeded whitespace changes>
$ pushd jdk/test/demo/zipfs
$ patch -Np7 < /tmp/test1.patch
$ popd
```

**Comment by apavlyutkin** [comment]
> Sorry for the delay on sorting this one out. It seems to have been bounced across all three repositories during the Git transition.
> 
> It now looks correct to me from a technical perspective. However, I am seeing some very odd whitespace differences between the two new test files when compared with the versions in 11u (either from the patch or the 11u-dev repository) e.g
> 
> ```diff
> -  @AfterMethod
> +    @AfterMethod
>      public void tearDown() throws IOException {
> -    deleteFiles();
> -  }
> +        deleteFiles();
> +    }
> ```
> 
> I don't know how this has occurred - I don't recall it being present in the original review - but I think the simplest way to fix it would be to store a diff between your 8u versions and the 11u versions using `-b`, replace the 8u ones with the ones from 11u, and then reapply your changes for 8u from the diff i.e. something like
> 
> ```shell
> $ diff -ub ../jdk11u-dev/test/jdk/jdk/nio/zipfs/ZipFSOutputStreamTest.java jdk/test/demo/zipfs/ZipFSOutputStreamTest.java  > /tmp/test1.patch
> $ cp ../jdk11u-dev/test/jdk/jdk/nio/zipfs/ZipFSOutputStreamTest.java jdk/test/demo/zipfs
> $ <edit test1.patch to remove any remaining unneeded whitespace changes>
> $ patch -Np0 < /tmp/test1.patch
> ```

fixed

File: jdk/test/demo/zipfs/LargeCompressedEntrySizeTest.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        try (FileSystem fs = FileSystems.newFileSystem(uri, Collections.singletonMap("create", "true"))) {
+            try (OutputStream os = Files.newOutputStream(fs.getPath(LARGE_FILE_NAME))) {
+                long remaining = largeEntrySize;
+                // create a chunk of random bytes which we keep writing out
+                final int chunkSize = 102400;
+                final byte[] chunk = new byte[chunkSize];
+                new Random().nextBytes(chunk);
+                final long start = System.currentTimeMillis();
+                for (long l = 0; l < largeEntrySize; l += chunkSize) {
+                    final int numToWrite = (int)Math.min(remaining, chunkSize);
```
Comment (gnu-andrew) [inline jdk/test/demo/zipfs/LargeCompressedEntrySizeTest.java:92]: 11u has a space between `(int)` and `Math`.

File: jdk/test/demo/zipfs/ZipFSOutputStreamTest.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            for (final Map.Entry<String, Long> entry : ZIP_ENTRIES.entrySet()) {
+                final Path entryPath = zipfs.getPath(entry.getKey());
+                try (final InputStream is = Files.newInputStream(entryPath)) {
+                    final byte[] buf = new byte[chunk.length];
+                    int numRead;
+                    long totalRead = 0;
+                    while ((numRead = is.read(buf)) != -1) {
+                        totalRead += numRead;
+                        // verify the content
+                        for (int i = 0, chunkoffset = (int)((totalRead - numRead) % chunk.length);
```
Comment (gnu-andrew) [inline jdk/test/demo/zipfs/ZipFSOutputStreamTest.java:120]: Again, 11u has a space between `(int)` and `((totalRead`

File: jdk/test/demo/zipfs/ZipFSOutputStreamTest.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * Repeatedly writes out to the outputstream, the chunk of data, till the number of bytes
+     * written to the stream equals the totalSize
+     */
+    private static void writeAsChunks(final OutputStream os, final byte[] chunk,
+                                      final long totalSize) throws IOException {
+        long remaining = totalSize;
+        for (long l = 0; l < totalSize; l += chunk.length) {
+            final int numToWrite = (int)Math.min(remaining, chunk.length);
```
Comment (gnu-andrew) [inline jdk/test/demo/zipfs/ZipFSOutputStreamTest.java:141]: 11u has a space between `(int)` and `Math`

**Review by gnu-andrew — COMMENTED** [review]
This version looks a lot better. There do still seem to be a few odd differences that have crept in, when compared to the patch in 11u, which I've noted in my review. These should be quick to fix, then we can get this in.

**Review by gnu-andrew — APPROVED** [review]
Brilliant, let's finally get this in.

**Comment by gnu-andrew** [comment]
I've added `jdk8u-fix-yes` to the bug.

**Pull request closed without merging**